Fix value omission false positive in `Style/MethodCallWithArgsParentheses`

gsamokovarov opened the pull request:
Omitting the parentheses in cases like the example below is ambigious, we don't need to enforce their omission in the `EnforcedStyle: omit_parentheses`

```ruby
var =
  unless object.action(value:, other:)
    condition || other_condition
  end
```

If the parentheses in the condition above are omitted, the semantic of
the code is changed:

```ruby
var =
  unless object.action value:, other:
    condition || other_condition
  end

# is interpreted as:

var =
  unless object.action value:, other: condition || other_condition
    # empty branch
  end
```

We should allow parentheses in value omissions in conditional nodes.

bbatsov commented:
I think all such cases where the parentheses are required should also be listed somewhere in the documentation examples.

gsamokovarov commented:
Technically, the parentheses are not required here but omitting them changes the behaviour of the code. There is ambiguity, which should be left to the developer to decide. I found this case by a crash, that I fix in https://github.com/rubocop/rubocop/pull/11712.

Documentation about all those cases can be useful so I will write some. Should it be in this PR as another commit?

bbatsov commented:
Another commit in this PR would be fine. 

> Technically, the parentheses are not required here but omitting them changes the behaviour of the code. There is ambiguity, which should be left to the developer to decide. I found this case by a crash that I fix in https://github.com/rubocop/rubocop/pull/11712.

Yeah, yeah - I get this. This particular case should be covered by some generic comment that we don't require/omit parentheses if this is going to change the meaning of the code. Something like a guiding principle for the cop in general.

gsamokovarov commented:
<img width="1369" alt="image" src="https://user-images.githubusercontent.com/604618/226404035-ac244fa3-00b8-4474-8b01-95608ffe1861.png">

I have also cleaned up the examples and added more examples where parentheses are required.

gsamokovarov commented on lib/rubocop/cop/style/method_call_with_args_parentheses.rb:
```diff
@@ -40,14 +38,16 @@ module Style
-      # because of the following issue: https://bugs.ruby-lang.org/issues/18396.
+      # NOTE: The style of `omit_parentheses` still allows parentheses in cases
+      # where omitting them results in ambiguous or syntactically incorrect
+      # code. For example, parentheses are required around a method with
```
Should I be exhaustive, though? Those are just a few cases, I can try and present all(most) of the cases here.

bbatsov replied:
I'm on the fence about this. Given how often we find problems with this cop I'm leaning towards more documentation, so it's clearer how it's supposed to work.

gsamokovarov replied:
<img width="1642" alt="image" src="https://user-images.githubusercontent.com/604618/226560071-de8759f9-7a54-4cb5-8179-80d12b9156d2.png">

☝️?
